test/integration/spokeagent_recovery_test.go

```diff
@@ -0,0 +1,204 @@
+			}
+			err := agentOptions.RunSpokeAgent(context.Background(), &controllercmd.ControllerContext{
+				KubeConfig:    spokeCfg,
+				EventRecorder: util.NewIntegrationTestEventRecorder("bootstrap-recoverytest"),
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		// wait the agent run a few seconds
+		time.Sleep(time.Second * 2)
+
+		// the spokecluster should not be created
+		_, err = util.GetSpokeCluster(clusterClient, spokeClusterName)
+		gomega.Expect(err).To(gomega.HaveOccurred())
+		gomega.Expect(errors.IsNotFound(err)).Should(gomega.BeTrue())
+
+		// the csr should not be created
+		_, err = util.FindUnapprovedSpokeCSR(kubeClient, spokeClusterName)
+		gomega.Expect(err).To(gomega.HaveOccurred())
+
+		// recover the invalid bootstrap kubeconfig file
+		err = util.CreateBootstrapKubeConfigWithCertAge(bootstrapFile, securePort, 24*time.Hour)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// the csr should be created after the bootstrap kubeconfig was recovered
+		gomega.Eventually(func() bool {
+			if _, err := util.FindUnapprovedSpokeCSR(kubeClient, spokeClusterName); err != nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// the spoke cluster should be created after the bootstrap kubeconfig was recovered
+		gomega.Eventually(func() bool {
+			if _, err := util.GetSpokeCluster(clusterClient, spokeClusterName); err != nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// simulate hub cluster admin accept the spoke cluster and approve the csr
+		err = util.AcceptSpokeCluster(clusterClient, spokeClusterName)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		err = util.ApproveSpokeClusterCSR(kubeClient, spokeClusterName, time.Hour*24)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// the hub kubeconfig secret should be filled after the csr is approved
+		gomega.Eventually(func() bool {
+			if _, err := util.GetFilledHubKubeConfigSecret(kubeClient, testNamespace, hubKubeconfigSecret); err != nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// simulate k8s to mount the hub kubeconfig secret
+		err = util.MountHubKubeConfigs(kubeClient, hubKubeconfigDir, testNamespace, hubKubeconfigSecret)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// the spoke cluster should have joined condition finally
+		gomega.Eventually(func() bool {
+			spokeCluster, err := util.GetSpokeCluster(clusterClient, spokeClusterName)
+			if err != nil {
+				return false
+			}
+			joined := helpers.FindSpokeClusterCondition(spokeCluster.Status.Conditions, clusterv1.SpokeClusterConditionJoined)
+			if joined == nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+	})
+
+	ginkgo.It("agent recovery from invalid hub kubeconfig", func() {
```

Comment:
I got a failure for this one locally:

```
Agent Recovery                                                                                                                                                                                                     
  agent recovery from invalid hub kubeconfig                                                                                                                                                                         /home/pmorie/go/src/github.com/open-cluster-management/registration/test/integration/spokeagent_recovery_test.go:112
I0518 17:22:08.911417   12134 spokeagent.go:88] Cluster name is "hubkubeconfig-recoverytest-spokecluster" and agent name is "zdppl"                                                                                I0518 17:22:08.915979   12134 shared_informer.go:223] Waiting for caches to sync for SpokeClusterCreatingController
I0518 17:22:08.916011   12134 shared_informer.go:230] Caches are synced for SpokeClusterCreatingController                                                                                                         I0518 17:22:08.916032   12134 base_controller.go:54] Starting #1 worker of SpokeClusterCreatingController controller ...                                         
I0518 17:22:08.916235   12134 spokeagent.go:160] Waiting for hub client config and spoke cluster to be ready                                                                                                       
I0518 17:22:08.918453   12134 shared_informer.go:223] Waiting for caches to sync for BootstrapClientCertForHubController                                                              
Event: [hubkubeconfig-recoverytest-spokeclustercreatingcontroller] SpokeClusterCreated: Spoke cluster "hubkubeconfig-recoverytest-spokecluster" created on hub 
I0518 17:22:09.018944   12134 shared_informer.go:230] Caches are synced for BootstrapClientCertForHubController 
I0518 17:22:09.019027   12134 base_controller.go:54] Starting #1 worker of BootstrapClientCertForHubController controller ...
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] NoValidCertificateFound: No valid client certificate for hub is found. Bootstrap is required 
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] CSRCreated: A csr "hubkubeconfig-recoverytest-spokecluster-cvvbs" is created 
Event: [hub-spokeclustercontroller] ClusterRoleCreated: Created ClusterRole.rbac.authorization.k8s.io/system:open-cluster-management:spokecluster:hubkubeconfig-recoverytest-spokecluster because it was missing 
Event: [hub-spokeclustercontroller] ClusterRoleBindingCreated: Created ClusterRoleBinding.rbac.authorization.k8s.io/system:open-cluster-management:spokecluster:hubkubeconfig-recoverytest-spokecluster because it 
was missing 
Event: [hub-spokeclustercontroller] NamespaceCreated: Created Namespace/hubkubeconfig-recoverytest-spokecluster because it was missing 
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] ClientCertificateCreated: A new client certificate is available 
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] NoValidCertificateFound: No valid client certificate for hub is found. Bootstrap is required 
Event: [hub-spokeclustercontroller] RoleCreated: Created Role.rbac.authorization.k8s.io/hubkubeconfig-recoverytest-spokecluster -n hubkubeconfig-recoverytest-spokecluster because it was missing 
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] CSRCreated: A csr "hubkubeconfig-recoverytest-spokecluster-zlc4p" is created 
Event: [hub-spokeclustercontroller] RoleBindingCreated: Created RoleBinding.rbac.authorization.k8s.io/hubkubeconfig-recoverytest-spokecluster -n hubkubeconfig-recoverytest-spokecluster because it was missing 
Event: [hub-spoke-cluster-controller] SpokeClusterAccepted: spoke cluster hubkubeconfig-recoverytest-spokecluster is accepted by hub cluster admin 
Event: [hubkubeconfig-recoverytest-bootstrapclientcertforhubcontroller] ClientCertificateCreated: A new client certificate is available 

• Failure [32.088 seconds]
Agent Recovery
/home/pmorie/go/src/github.com/open-cluster-management/registration/test/integration/spokeagent_recovery_test.go:18
  agent recovery from invalid hub kubeconfig [It]
  /home/pmorie/go/src/github.com/open-cluster-management/registration/test/integration/spokeagent_recovery_test.go:112

  Timed out after 30.000s.
  Expected
      <bool>: false
  to be true

  /home/pmorie/go/src/github.com/open-cluster-management/registration/test/integration/spokeagent_recovery_test.go:202
------------------------------
STEP: tearing down the test environment
E0518 17:22:41.002078   12134 reflector.go:382] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to watch *v1beta1.CertificateSigningRequest: Get https://127.0.0.1:44663/apis/certificates.k$
s.io/v1beta1/certificatesigningrequests?allowWatchBookmarks=true&resourceVersion=115&timeout=9m19s&timeoutSeconds=559&watch=true: dial tcp 127.0.0.1:44663: connect: connection refused
E0518 17:22:41.003048   12134 reflector.go:382] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to watch *v1beta1.CertificateSigningRequest: Get https://127.0.0.1:44663/apis/certificates.k$
s.io/v1beta1/certificatesigningrequests?allowWatchBookmarks=true&resourceVersion=115&timeout=5m8s&timeoutSeconds=308&watch=true: dial tcp 127.0.0.1:44663: connect: connection refused
```

Reply:
I get flakes 15-20% of the time on this case

Reply:
I run this many times in my local env (4 groups, 10 times per group), but all of them are passed, from code, I think it may be failed on `stopBootstrap()` but I'm not sure why it happens in your local env ...

```
bootstrapCtx, stopBootstrap := context.WithCancel(ctx)
...

if err := wait.PollImmediateInfinite(1*time.Second, o.hasValidHubClientConfig); 
...
stopBootstrap() 
```